webhook/workload/mutating/workload_update_handler.go

```diff
@@ -0,0 +1,258 @@
+		}
+		if !changed {
+			return admission.Allowed("")
+		}
+		marshalled, err := json.Marshal(newObj)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		return admission.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, marshalled)
+	// native k8s deloyment
+	case apps.SchemeGroupVersion.Group:
+		if req.Kind.Kind != util.ControllerKindDep.Kind {
+			return admission.Allowed("")
+		}
+		// check deployment
+		newObj := &apps.Deployment{}
+		if err := h.Decoder.Decode(req, newObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		oldObj := &apps.Deployment{}
+		if err := h.Decoder.Decode(
+			admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}},
+			oldObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		err, changed := h.handlerDeployment(newObj, oldObj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if !changed {
+			return admission.Allowed("")
+		}
+		marshalled, err := json.Marshal(newObj)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		return admission.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, marshalled)
+	}
+	return admission.Allowed("")
+}
+
+func (h *WorkloadHandler) handlerDeployment(newObj, oldObj *apps.Deployment) (err error, changed bool) {
+	// in rollout progressing
+	if state, _ := util.GetRolloutState(newObj.Annotations); state != nil {
+		// deployment paused=false is not allowed until the rollout is completed
+		if newObj.Spec.Paused == false {
+			changed = true
+			newObj.Spec.Paused = true
+			klog.Warningf("deployment(%s/%s) is in rollout(%s) progressing, and set paused=true", newObj.Namespace, newObj.Name, state.RolloutName)
+		}
+		return
+	}
+
+	// indicate whether the workload can enter the rollout process
+	// 1. replicas > 0
+	if newObj.Spec.Replicas != nil && *newObj.Spec.Replicas == 0 {
+		return
+	}
+	// 2. deployment.spec.strategy.type must be RollingUpdate
+	if newObj.Spec.Strategy.Type == apps.RecreateDeploymentStrategyType {
+		klog.Warningf("deployment(%s/%s) strategy type is 'Recreate', rollout will not work on it", newObj.Namespace, newObj.Name)
+		return
+	}
+	// 3. deployment.spec.PodTemplate not change
+	if util.EqualIgnoreHash(&oldObj.Spec.Template, &newObj.Spec.Template) {
+		return
+	}
+	// 4. the deployment must be in a stable version (only one version of rs)
+	stableRs, err := h.Finder.GetDeploymentStableRs(newObj)
+	if err != nil {
+		return
+	} else if stableRs == nil {
+		return
+	}
+	// 5. have matched rollout crd
+	rollout, err := h.fetchMatchedRollout(newObj)
+	if err != nil {
+		return
+	} else if rollout == nil {
+		return
+	}
+	klog.Infof("deployment(%s/%s) will be in rollout progressing, and set paused=true", newObj.Namespace, newObj.Name)
+
+	changed = true
+	// need set workload paused = true
+	newObj.Spec.Paused = true
```

Comment:
what if we finish the canary phrase and require the rolling update of the deployment?  In such case, we do need set the Paused to false right?